diagnostics/ngtf_graph_viewer.py

```diff
@@ -0,0 +1,179 @@
+    return gdef
+
+
+def graphdef_to_dot(gdef, dot_output):
+    with open(dot_output, "wb") as f:
+        print("digraph graphname {", file=f)
+        for node in gdef.node:
+            output_name = node.name
+            print("  \"" + output_name + "\" [label=\"" + node.op + "\"];", file=f)
+            for input_full_name in node.input:
+                parts = input_full_name.split(":")
+                input_name = re.sub(r"^\^", "", parts[0])
+                print("  \"" + input_name + "\" -> \"" + output_name + "\";", file=f)
+        print("}", file=f)
+    print("\n"+('=-'*30))
+    print("Created DOT file '" + dot_output + "'.")
+    print("Can be converted to pdf using: dot -Tpdf " +
+          dot_output+" -o "+dot_output+".pdf")
+    print('=-'*30)
+
+
+def protobuf_to_dot(input_filename, dot_dir, input_binary=False, node_map={}):
+    gdef = preprocess(input_filename, dot_dir, input_binary, node_map)
+    graphdef_to_dot(gdef, dot_dir.rstrip('/') + '/' +
+                    os.path.basename(input_filename) + '.dot')
+
+
+def graphdef_to_tensorboard(gdef, tensorboard_output):
+    # convert graphdef to graph, even though FileWriter can accepts graphdefs.
+    # this is because FileWriter has deprecated graphdef as inputs, and prefers graphs as inputs
+    with tf.Session() as sess:
+        tf.import_graph_def(gdef)
+        writer = tf.summary.FileWriter(tensorboard_output, sess.graph)
+        # TODO: try with tf master
+        # wont work now if we have NGraphVariable, NGraphEncapsulateOp
+        # TODO: How about supporting NGraphVariable and NGraphEncapsulateOp by switching their optype with something TB knows
+        writer.flush()
+        writer.close()
+    # It seems NGraphVariable and NGraphEncapsulateOp are registered in C++ but not in python
+    print("\n"+('=-'*30)+"\nTo view Tensorboard:")
+    print("1) Run this command: tensorboard --logdir " + tensorboard_output)
+    print("2) Go to the URL it provides or http://localhost:6006/\n"+('=-'*30)+"\n")
+
+
+def protobuf_to_grouped_tensorboard(input_filename, tensorboard_dir, input_binary=False, node_map={}):
+    gdef = preprocess(input_filename, tensorboard_dir, input_binary, node_map)
+    graphdef_to_tensorboard(gdef, tensorboard_dir)
+
+
+visualizations_supported = [protobuf_to_dot, protobuf_to_grouped_tensorboard]
+
+if __name__ == "__main__":
+    helptxt = '''
+    Convert protobuf to different visualizations (dot, tensorboard).
+
+    Sample usage from command line:
+    python protobuf_visualize.py pbtxtfile.pbtxt ./vis  # read pbtxt and generate TB
+    python protobuf_visualize.py -v 1 pbtxtfile.pbtxt ./vis  # read pbtxt and generate dot
+    python protobuf_visualize.py -b pbtxtfile.pb ./vis  # read pb and generate TB
+    python protobuf_visualize.py -b -v 1 pbtxtfile.pb ./vis  # read pb and generate dot
+    python protobuf_visualize.py -c nodemap.pkl pbtxtfile.pbtxt ./vis  # read pbtxt, remap node names and generate TB
```

Comment:
1. this is auto pepped. But that long line persists because it is a multiline string.

2. This is the option that we use for rendering encapsulates on TB. The way to use case is at the end of the README